[inductor] Fix inputs with existing offsets

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108169
* #108172
* __->__ #108168

This cherrypicks the reinterpret_tensor change from #102625 in order to fix a subtle correctness bug when the graph inputs already have a storage_offset set.

The view change also fixes some issues with quantized models in torchbench.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

jansel committed:
[inductor] Fix inputs with existing offsets

[ghstack-poisoned]

desertfire commented on torch/csrc/inductor/inductor_ops.cpp:
```diff
@@ -22,10 +22,29 @@ Tensor _mm_plus_mm(
   return out;
 }
 
+// Similar to as_strided with the following differences
```
Probably worth to also mention these in the `ir.ReinterpretView` class's comment.

desertfire commented on torch/_inductor/ir.py:
```diff
@@ -1873,12 +1873,9 @@ def codegen_reference(self):
-                f"{namespace}as_strided({self.get_name()}, {size}, {stride}, {offset})"
-            )
-        return f"{namespace}as_strided({self.get_name()}, {size}, {stride})"
+            return f"reinterpret_tensor({self.get_name()}, {size}, {stride}, {offset})"
```
Nit: we don't really need the conditional checking here?

jansel committed:
Update on "[inductor] Fix inputs with existing offsets"


This cherrypicks the reinterpret_tensor change from #102625 in order to fix a subtle correctness bug when the graph inputs already have a storage_offset set.

The view change also fixes some issues with quantized models in torchbench.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jansel commented:
@pytorchbot merge
